### ISSUE TITLE
Method call compiles with javac but not ecj

### ISSUE THREAD

**Issue by coehlrich** [eclipse-jdt/eclipse.jdt.core#3457]
```javac
public class Test {
	public void test() {
		this.error(new TypeToken<A2<?>>() {});
	}
	
	public <T extends B1> void error(TypeToken<? extends A1<? extends T>> type) {}
	public static abstract class TypeToken<T> {}
	public static class A1<T extends B1> {}
	public static class A2<T extends B2> extends A1<T> {}
	
	public static class B1 {}
	public static class B2 extends B1 {}
}
```
javac compiles successfully while ecj fails with the error `The method error(Test.TypeToken<? extends Test.A1<? extends T>>) in the type Test is not applicable for the arguments (new Test.TypeToken<Test.A2<?>>(){})`

The issue seems to be that a type bound of `T#0 :> java.lang.Object` gets added from the constraint that `A2<?>` is a subtype of `A1<? extends T#0>` due to the unbounded wildcard from `A2<?>` being preserved while getting the supertype even though [JLS §4.10.2](https://docs.oracle.com/javase/specs/jls/se21/html/jls-4.html#jls-4.10.2) suggests that the supertypes of the capture conversion of `A2<?>` should be used instead

> Given a generic class or interface C with type parameters F1,...,Fn (n > 0), the direct supertypes of the parameterized type C<R1,...,Rn> where at least one of the Ri (1 ≤ i ≤ n) is a wildcard type argument, are the direct supertypes of the parameterized type C<X1,...,Xn> which is the result of applying capture conversion to C<R1,...,Rn> ([§5.1.10](https://docs.oracle.com/javase/specs/jls/se21/html/jls-5.html#jls-5.1.10)).

**Comment by srikanth-sankaran** [comment]
> @srikanth-sankaran It would be great to get your thoughts, since the final pass from https://bugs.eclipse.org/426537 was your invention :

Gee, I am sure I remember absolutely nothing about it - I wouldn't take my word on the matter as counting for much :-(

**Comment by stephan-herrmann** [comment]
With experimentally sprinkling scopes and capturing into "relevant" locations, I still fail to see compatibility at
* `<anonType>`.isCompatibleWith(`TypeToken<? extends A1<? extends capture#1-of ?>>`) is `false`
* because `TypeToken<? extends A1<? extends capture#1-of ?>>`.isSuperclassOf(`<anonType>`) is `false`
* because `TypeToken<A2<?>>`.isEquivalent(`TypeToken<? extends A1<? extends capture#1-of ?>>`) is `false`
   * using `TypeToken<A2<?>>` as the superclass of the anon type
* because `A2<?>`.isTypeArgumentedContainedBy(`? extends A1<? extends capture#1-of ?>`) is `false`
* because`A2<?>`.isCompatibleWith(`A1<? extends capture#1-of ?>`) is `false`
* because `A1<? extends capture#1-of ?>`.isSuperclassOf(`A2<?>`) is `false`
* because `A1<capture#2-of ?>`.isEquivalentTo(`A1<? extends capture#1-of ?>`) is `false`
   * using `A1<capture#2-of ?>` as the superclass of `A2<?>`
* because `capture#2-of ?`.isTypeArgumentedContainedBy(`? extends capture#1-of ?`) is `false`
* because `capture#2-of ?`.isCompatibleWith(`capture#1-of ?`) is `false`

Could we reasonably unify both captures (they share the same wildcard)?
* semantically?
* technically?

**Comment by stephan-herrmann** [comment]
The initial fix caused a regression, viz https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/2981

Reduced example:
```java
import java.util.ArrayList;
import java.util.Collection;

interface IQuery<T> { }

public class QueryUtil {
	public static <T> IQuery<T> createCompoundQuery(IQuery<? extends T> query1, IQuery<T> query2, boolean and) {
		ArrayList<IQuery<? extends T>> queries = new ArrayList<>(2);
		queries.add(query1);
		queries.add(query2);
		return createCompoundQuery(queries, and);
	}
	public static <T> IQuery<T> createCompoundQuery(Collection<? extends IQuery<? extends T>> queries, boolean and) {
		return null;
	}
}
```

Again failure happens in the "final pass over the initial constraints", where now we have this constraint after applying the substitution: 
* `⟨queries → Collection<? extends IQuery<? extends capture#2-of ? extends T>>⟩`

Ultimately we see the following violated:
* `? extends T <= ? extends capture#2-of ? extends T`

It's that intermediate capture that is introduced by the existing "fix".

In summary:
* #3910 aligned us with [JLS §4.10.2](https://docs.oracle.com/javase/specs/jls/se21/html/jls-4.html#jls-4.10.2)  (partially: only during type inference)
* in #3915 I had to compensate in a situation for which I could not find any backing by JLS (in particular, I am not even sure if a type of shape `? extends capture#2-of ? extends T` is well-formed to begin with. I claim, though, that this compensation is exactly in the spirit of https://bugs.openjdk.org/browse/JDK-8016207: if javac widens captures at unspecified times, we should certainly do similar things. With this, the second fix is opportunistic by definition and necessity.
